src/ray/util/event.cc

```diff
@@ -16,11 +16,215 @@
+      return;
+    }
+  }
+
+  if (!stop_writing_) {
+    errno = 0;
+    // use message_len + 1 because LogSink the last char is '\n' "
+    fwrite(message, 1, message_len, file_);
+    if (errno == ENOSPC) {
+      stop_writing_ = true;
+      return;
+    } else {
+      file_length_ += message_len;
+      bytes_since_flush_ += message_len;
+    }
+  }
+
+  if (force_flush || bytes_since_flush_ > max_flush_size_) {
+    FlushUnlocked();
+  }
+}
+
+///
+/// LogEventReporter
+///
+LogEventReporter::LogEventReporter(rpc::Event_SourceType source_type, std::string log_dir)
+    : log_dir_(log_dir) {
+  RAY_CHECK(log_dir_ != "");
+  if (log_dir_.back() != '/') {
+    log_dir_ += '/';
+  }
+  boost::filesystem::create_directories(log_dir_);
+  file_name_ = "event_" + Event_SourceType_Name(source_type);
+  if (source_type == rpc::Event_SourceType::Event_SourceType_CORE_WORKER ||
+      source_type == rpc::Event_SourceType::Event_SourceType_COMMON) {
+    file_name_ += "_" + std::to_string(getpid());
+  }
+
+  log_sink_ = std::unique_ptr<LogFileObject>(new LogFileObject(log_dir_, file_name_));
+}
+
+std::string LogEventReporter::EventToString(const rpc::Event &event) {
+  std::stringstream result;
+
+  result << "{\"time_stamp\":\"" << event.timestamp() << "\",\"severity\":\""
+         << Event_Severity_Name(event.severity()) << "\",\"label\":\"" << event.label()
+         << "\",\"event_id\":\"" << event.event_id() << "\",\"source_type\":\""
+         << Event_SourceType_Name(event.source_type()) << "\",\"host_name\":\""
+         << event.source_hostname() << "\",\"pid\":\""
+         << std::to_string(event.source_pid());
+
+  result << "\",\"message\":\"";
+
+  const std::string &message = event.message();
+  for (int i = 0, len = message.size(); i < len; ++i) {
+    if (message[i] == '\n' || message[i] == '\r') {
+      result << "\\n";
+    } else if (message[i] == '\"') {
+      result << "\'";
+    } else {
+      result << message[i];
+    }
+  }
+
+  result << "\",\"custom_fields\":{";
+  int count = 0;
+  for (auto &ele : event.custom_fields()) {
+    if (count) result << ",";
+    result << "\"" << ele.first << "\":\"" << ele.second << "\"";
+    count++;
+  }
+
+  result << "}}";
+
+  return result.str();
+}
+
+void LogEventReporter::Report(const rpc::Event &event) {
+  RAY_CHECK(Event_SourceType_IsValid(event.source_type()));
+  RAY_CHECK(Event_Severity_IsValid(event.severity()));
+  std::string result = EventToString(event) + "\n";
+  log_sink_->Write(
+      result.c_str(), result.size(),
+      event.severity() == rpc::Event_Severity::Event_Severity_FATAL ? true : false);
+}
+///
 /// EventManager
 ///
+
+std::unique_ptr<EventManager> EventManager::instance_ = nullptr;
+absl::Mutex EventManager::mutex_;
+
 EventManager &EventManager::Instance() {
-  static EventManager instance_;
-  return instance_;
+  if (instance_ == nullptr) {
+    absl::MutexLock lock(&mutex_);
+    if (instance_ == nullptr) {
```

Comment:
It seems duplicated null-check?

Reply:
Here to prevent multi-thread calling Instance API. I have to guarantee the event manager is constructed only once.

Reply:
So why not add lock_gurad outside

Reply:
Obviously it is more efficient. If the instance is already initialized, we don't need to add lock.

Reply:
It looks so weird. You'd better make sure it's lock-free if we should use double-check here.